numpy/core/src/multiarray/convert.c

```diff
@@ -302,57 +302,34 @@ PyArray_ToString(PyArrayObject *self, NPY_ORDER order)
     npy_intp i;
     char *dptr;
     int elsize;
-    PyObject *ret;
+    PyObject *ravel_out, *ret;
     PyArrayIterObject *it;
 
-    if (order == NPY_ANYORDER)
-        order = PyArray_ISFORTRAN(self);
-
-    /*        if (PyArray_TYPE(self) == NPY_OBJECT) {
-              PyErr_SetString(PyExc_ValueError, "a string for the data" \
-              "in an object array is not appropriate");
-              return NULL;
-              }
-    */
-
     numbytes = PyArray_NBYTES(self);
-    if ((PyArray_IS_C_CONTIGUOUS(self) && (order == NPY_CORDER))
-        || (PyArray_IS_F_CONTIGUOUS(self) && (order == NPY_FORTRANORDER))) {
-        ret = PyBytes_FromStringAndSize(PyArray_DATA(self), (Py_ssize_t) numbytes);
+
+    ravel_out = PyArray_Ravel(self, order);
+    it = (PyArrayIterObject *)PyArray_IterNew(ravel_out);
+    if (ravel_out == NULL) {
+        Py_DECREF(it);
+        return NULL;
     }
-    else {
-        PyObject *new;
-        if (order == NPY_FORTRANORDER) {
-            /* iterators are always in C-order */
-            new = PyArray_Transpose(self, NULL);
```

Comment:
> Calling ravel will work, but may create an additional copy unfortunately (I somewhat expected we do that anyway, but we don't...), its just one copy more though, maybe we should not care too much...

wouldnt this transpose also create a copy, or are you saying that we avoid the copy for all other cases in the else earlier which we would now not be avoiding ?

Reply:
Transpose never actually copies anything.

Reply:
okay thanks a lot for your explanations @seberg . I will try and proceed with the iterator approach that you outlined above. 

also

> Or maybe we just put the axis-reorganizing ravel does into an internal helper, so that we do not have to duplicate it.

this approach will still create a copy right ?

Reply:
Yeah, its a bit tricky. I suppose it should be possible to do something like this:
```
string = empty_string
data = string.data_ptr

self_transposed = self.transpose(optimal_order)
new_array = np.ndarray(data, self_transposed.shape)  # PyArray_NewFromDescr

np.copyto(new_array, self_transposed)
```